### PR TITLE
Query method is immutable reference

### LINKED ISSUE #16: query method requires a mutable reference to Index
The ```query``` API requires you to pass a mutable reference to the index. The reason for this is that vacuuming might occur (complete deletion of removed documents) which is a mutation. It would be more ideal if the the ```query``` would just take a reference to the index instead, yet still, if possible, use the trie index walk during query as an opportunity to vacuum (or as an opportunity in finding out where to vacuum later).

Based on discussions in #15

### PR DESCRIPTION
**Pull request by tmpfs**
Closes #16.

### CODE REVIEW

**Comment by tmpfs** [comment]
You were right @marcus-pousette, we got another small perf improvement 👍

**Comment by tmpfs** [comment]
@marcus-pousette, I had to revert the changes I made to the `Filter` type to accommodate owned values which is required for when filters need to perform transformations (such as `to_lowercase()`). Using a `std::borrow::Cow` we can allow for no-op filters that just do `Cow::from(s)` and also transformations that incur allocations such as `Cow::Owned(s.to_lowercase())`. Some of those performance gains in `add_document()` have been lost as we have to use `String` instead of `&str` but it does allow for no-op filters to have minimal overhead.

File: src/query.rs [hunk 1]
```diff
@@ -55,52 +44,49 @@ impl<T: Eq + Hash + Copy + Debug> Index<T> {
                         &self.arena_index,
                     );
                     if let Some(term_node_index) = term_node_option {
-                        let document_frequency =
-                            self.disconnect_and_count_documents(term_node_index, removed);
+                        let document_frequency = self.count_documents(term_node_index);
                         let term_node = self.arena_index.get(term_node_index).unwrap();
                         if let Some(term_node_option_first_doc) = term_node.first_doc {
-                            if document_frequency > 0 {
```
Comment (marcus-pousette) [inline src/query.rs:62]: Why was this removed?  The reason for this if statement is a purely a performance once. It does not make sense to process the term any longer if it does not mach any documents

File: src/index.rs [hunk 2]
```diff
@@ -278,6 +286,18 @@ impl<T: Eq + Hash + Copy + Debug> Index<T> {
...
     }
 
+    /// Count the document frequency.
+    pub(crate) fn count_documents(&self, node_index: ArenaIndex<InvertedIndexNode<T>>) -> usize {
+        let node = self.arena_index.get(node_index).unwrap();
+        let mut pointer_option = node.first_doc;
+        let mut document_frequency = 0;
+        while let Some(pointer) = pointer_option {
+            document_frequency += 1;
+            pointer_option = self.arena_doc.get(pointer).unwrap().next;
```
Comment (marcus-pousette) [inline src/index.rs:296]: Perhaps, we should still check whether the document is removed before bumping the document_frequency? 

```suggestion
            while let Some(pointer) = pointer_option {
                let is_removed = match removed {
                    Some(set)  => {
                        if set.contains(&self.arena_doc.get(pointer).unwrap().details_key)
                        {
                            match &prev_pointer {
                                None => {
                                    node.first_doc = self.arena_doc.get(pointer).unwrap().next;
                                }
                                Some(prev) => {
                                    self.arena_doc.get_mut(*prev).unwrap().next =
                                self.arena_doc.get(pointer).unwrap().next;
                                }
                            }
                        }
                        true
                    },
                    None => false
                };
                if !is_removed
                {
                    document_frequency += 1;
                    prev_pointer = Some(pointer);
                }
                pointer_option = self.arena_doc.get(pointer).unwrap().next;
                
            }
```

I curious how this could pass without any test errors

File: src/index.rs [hunk 3]
```diff
@@ -26,27 +26,19 @@ pub struct Index<T> {
 
     pub(crate) arena_index: StandardArena<InvertedIndexNode<T>>,
     pub(crate) arena_doc: StandardArena<DocumentPointer<T>>,
+
+    /// Documents that have been removed from the index but
+    /// need to be purged.
+    removed: Option<HashSet<T>>,
```
Comment (marcus-pousette) [inline src/index.rs:32]: This is a great change

**Review by marcus-pousette — COMMENTED** [review]
Left some comments, that might be worth considering. 

but the main problem as that I think the ```count_documents``` fn has to respect ```removed``` set

Good find and solution regarding the ```Filter``` I thought about this when reviewing your last PR but I thought there would be a explicit lifetime trick you could do in order to circumvent this, but I realize now it is not possible yet to associate lifetimes with borrows in a ```fn```.

File: src/query.rs (hunk 1 above)
Comment (tmpfs) [reply src/query.rs:62]: My bad, that has been restored.

File: src/index.rs (hunk 2 above)
Comment (marcus-pousette) [reply src/index.rs:296]: Edit* I updated my suggestion, I copied too freely from master

File: src/index.rs [hunk 4]
```diff
@@ -98,12 +97,13 @@ impl<T: Eq + Hash + Copy + Debug> Index<T> {
...
                 // filter and count terms, ignore empty strings
                 let mut filtered_terms_count = 0;
-                for mut term in terms {
-                    term = filter(term);
+                for term in terms {
+                    let filtered = filter(term);
+                    let term = filtered.as_ref().to_owned();
                     if !term.is_empty() {
-                        all_terms.push(term);
+                        all_terms.push(term.clone());
```
[the remark this answers is not in the document]
Comment (marcus-pousette) [reply src/index.rs:104]: Perhaps a ```Rc``` could also work, here but then you are paying another kind of price

File: src/lib.rs [hunk 5]
```diff
@@ -12,12 +14,13 @@ pub type FieldAccessor<D> = fn(&D) -> Option<&str>;
 pub type Tokenizer = fn(&str) -> Vec<&str>;
```
[the remark this answers is not in the document]
Comment (tmpfs) [reply src/lib.rs]: Yep, thats done now 👍

File: src/index.rs [hunk 6]
```diff
@@ -103,17 +103,9 @@ impl<T: Eq + Hash + Copy + Debug> Index<T> {
...
-                                new_count[i] += 1;
-                                term_counts.insert(term, new_count);
-                            }
-                            Some(c) => {
-                                c[i] += 1;
-                            }
-                        }
+                        let counts = term_counts.entry(term)
+                            .or_insert(vec![0; fields_len]);
+                        counts[i] += 1;
```
Comment (marcus-pousette) [inline src/index.rs:109]: Nice!

File: src/index.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (tmpfs) [reply src/index.rs:104]: I think using `Cow` here over-complicates this code, I gave it a try and couldn't keep the borrow checker happy so maybe just leave it as is for now? We can revisit later.

File: src/index.rs (hunk 2 above)
Comment (tmpfs) [reply src/index.rs]: Good catch, this should be fixed now 👍

File: src/index.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (marcus-pousette) [reply src/index.rs:104]: Sounds good!

Yes, it is a bit messy. I got a working solution locally. I could just submit it after this PR is merged.

**Review by marcus-pousette — COMMENTED** [review]
I think this looks good now. Have you done all the changes you wanted to do?

File: src/index.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (tmpfs) [reply src/index.rs:104]: Ok cool, nice one 👍

**Comment by tmpfs** [comment]
Sure @marcus-pousette, I think this is good enought for this PR, ping me when you push your `Cow` fix, interested to see how that shapes up!
